pyscriptjs/src/main.ts

```diff
@@ -1,31 +1,103 @@
 
-/* eslint-disable @typescript-eslint/no-unused-vars */
-const xPyScript = customElements.define('py-script', PyScript);
-const xPyLoader = customElements.define('py-loader', PyLoader);
-const xPyConfig = customElements.define('py-config', PyConfig);
-const xPyEnv = customElements.define('py-env', PyEnv);
-/* eslint-disable @typescript-eslint/no-unused-vars */
-
-// As first thing, loop for application configs
-logger.info('checking for py-config');
-const config: PyConfig = document.querySelector('py-config');
-if (!config) {
-    const loader = document.createElement('py-config');
-    document.body.append(loader);
+// XXX this should be killed eventually
+let runtimeSpec: Runtime;
+runtimeLoaded.subscribe(value => {
+    runtimeSpec = value;
+});
+
+
+class PyScriptApp {
+
+    config: AppConfig;
+
+    main() {
+        this.loadConfig();
+        this.initialize();
+
+        /* eslint-disable @typescript-eslint/no-unused-vars */
+        const xPyScript = customElements.define('py-script', PyScript);
+        const xPyLoader = customElements.define('py-loader', PyLoader);
+        const xPyEnv = customElements.define('py-env', PyEnv);
+        /* eslint-disable @typescript-eslint/no-unused-vars */
+
+        // add loader to the page body
+        logger.info('add py-loader');
+        const loader = <PyLoader>document.createElement('py-loader');
+        document.body.append(loader);
+        globalLoader.set(loader);
+    }
+
+    loadConfig() {
+        // find the <py-config> tag. If not found, we get null which means
+        // "use the default config"
+        // XXX: what happens if we have multiple ones?
+        logger.info('searching for <py-config>');
+        const el = document.querySelector('py-config');
+        this.config = loadConfigFromElement(el);
+        logger.info('config loaded:\n' + JSON.stringify(this.config, null, 2));
+    }
+
+    initialize() {
+        addInitializer(this.loadPackages);
+        addInitializer(this.loadPaths);
+        this.loadRuntimes();
+    }
+
+    loadPackages = async () => {
+        logger.info("Packages to install: ", this.config.packages);
+        await runtimeSpec.installPackage(this.config.packages);
+    }
+
+    loadPaths = async () => {
+        const paths = this.config.paths;
+        logger.info("Paths to load: ", paths)
+        for (const singleFile of paths) {
+            logger.info(`  loading path: ${singleFile}`);
+            try {
+                await runtimeSpec.loadFromFile(singleFile);
+            } catch (e) {
+                //Should we still export full error contents to console?
+                handleFetchError(<Error>e, singleFile);
+            }
+        }
+        logger.info("All paths loaded");
+    }
+
+    loadRuntimes() {
+        logger.info('Initializing runtimes');
+        for (const runtime of this.config.runtimes) {
+            const runtimeObj: Runtime = new PyodideRuntime(this.config, runtime.src,
```

Comment:
I agree. Personally, I would like to change the config to allow only one runtime instead of having a list-which-is-not-really-a-list-because-it-doesnt-work-if-you-have-more-than-one.
But as I said above, in this PR I tried to move around the logic without actually trying to change it too much.

Reply:
KISS (keep is simple, stupid)... totally agree. Also, if you need more than one runtime, you're probably doing it wrong.